### PR TITLE
📁 Export Catalog Modal & Selections

### PR DESCRIPTION
**Pull request by arifluthfi16**

### CODE REVIEW

File: kuwala/core/backend/app/main.py [hunk 1]
```diff
@@ -127,21 +137,21 @@ def populate_db():
     )
     transformation_catalog_categories = json.load(file)
 
-    for tcc in transformation_catalog_categories:
+    for export_cat in transformation_catalog_categories:
```
Comment (mattigrthr) [inline kuwala/core/backend/app/main.py]: This line shouldn't be changed.

File: kuwala/core/backend/app/main.py [hunk 2]
```diff
@@ -127,21 +137,21 @@ def populate_db():
...
         try:
             get_object_by_id(
                 db=db,
                 model=transformation_catalog_category_models.TransformationCatalogCategory,
-                object_id=tcc["id"],
+                object_id=export_cat["id"],
```
Comment (mattigrthr) [inline kuwala/core/backend/app/main.py]: This line shouldn't be changed.

File: kuwala/core/backend/app/main.py [hunk 3]
```diff
@@ -127,21 +137,21 @@ def populate_db():
...
             if e.status_code == 404:
                 create_transformation_catalog_category(
                     db=db,
                     transformation_catalog_category=transformation_catalog_category_schemas.TransformationCatalogCategoryCreate(
-                        id=tcc["id"],
-                        name=tcc["name"],
-                        icon=tcc["icon"],
+                        id=export_cat["id"],
+                        name=export_cat["name"],
+                        icon=export_cat["icon"],
```
Comment (mattigrthr) [inline kuwala/core/backend/app/main.py]: This line shouldn't be changed.

File: kuwala/core/backend/app/main.py [hunk 4]
```diff
@@ -159,36 +169,101 @@ def populate_db():
         )
     )
 
-    for tcc in transformation_catalog_categories:
-        transformations = list(map(lambda tr: f"{tcc}/{tr}", os.listdir(tcc)))
+    for export_cat in transformation_catalog_categories:
+        transformations = list(
+            map(lambda tr: f"{export_cat}/{tr}", os.listdir(export_cat))
+        )
```
Comment (mattigrthr) [inline kuwala/core/backend/app/main.py]: This line shouldn't be changed.

File: kuwala/core/backend/app/main.py [hunk 5]
```diff
@@ -159,36 +169,101 @@ def populate_db():
...
 
-        for t in transformations:
-            file = open(t)
-            t = json.load(file)
+        for ex in transformations:
+            file = open(ex)
+            ex = json.load(file)
```
Comment (mattigrthr) [inline kuwala/core/backend/app/main.py]: This line shouldn't be changed.

File: kuwala/core/backend/app/main.py [hunk 6]
```diff
@@ -159,36 +169,101 @@ def populate_db():
...
 
             try:
                 get_object_by_id(
                     db=db,
                     model=transformation_catalog_models.TransformationCatalogItem,
-                    object_id=t["id"],
+                    object_id=ex["id"],
```
Comment (mattigrthr) [inline kuwala/core/backend/app/main.py]: This line shouldn't be changed.

File: kuwala/core/backend/app/main.py [hunk 7]
```diff
@@ -159,36 +169,101 @@ def populate_db():
...
+                            description=ex["description"],
+                            required_column_types=ex["required_column_types"],
+                            optional_column_types=ex["optional_column_types"],
+                            min_number_of_input_blocks=ex["min_number_of_input_blocks"],
+                            max_number_of_input_blocks=ex["max_number_of_input_blocks"],
+                            macro_parameters=json.dumps(ex["macro_parameters"]),
+                            examples_before=json.dumps(ex["examples_before"]),
+                            examples_after=json.dumps(ex["examples_after"]),
+                        ),
+                    )
```
Comment (mattigrthr) [inline kuwala/core/backend/app/main.py]: This line shouldn't be changed.

File: kuwala/core/backend/app/main.py [hunk 8]
```diff
@@ -159,36 +169,101 @@ def populate_db():
...
+                    ),
+                )
+
+    # Add export catalog items into database
+    export_catalog_path = os.path.join(script_dir, "./resources/export_catalog")
+    export_catalog_categories = list(
+        filter(
+            lambda d: os.path.isdir(d),
+            map(
+                lambda d: f"{export_catalog_path}/{d}",
```
Comment (mattigrthr) [inline kuwala/core/backend/app/main.py]: Why do you name the variable `d`?

File: kuwala/core/backend/app/main.py [hunk 9]
```diff
@@ -159,36 +169,101 @@ def populate_db():
...
+                os.listdir(export_catalog_path),
+            ),
+        )
+    )
+
+    for export_cat in export_catalog_categories:
+        exports = list(map(lambda tr: f"{export_cat}/{tr}", os.listdir(export_cat)))
```
Comment (mattigrthr) [inline kuwala/core/backend/app/main.py]: Why do you name the variable `tr`?

File: kuwala/core/backend/app/resources/export_catalog/categories.json [hunk 10]
```diff
@@ -0,0 +1,12 @@
+[
+  {
+    "id": "file",
+    "name": "File",
+    "icon": "file-export"
+  },
+  {
+    "id": "bi_tool",
+    "name": "Bi Tool",
```
Comment (mattigrthr) [inline kuwala/core/backend/app/resources/export_catalog/categories.json]: Should be "BI Tool".

File: kuwala/core/backend/app/resources/export_catalog/file/download_as_csv.json [hunk 11]
```diff
@@ -0,0 +1,21 @@
+{
+  "id": "download_as_csv",
+  "category": "file",
+  "name": "Download as CSV",
+  "icon": "file-csv",
+  "description": "With this export, you can export the result of a block into a CSV File",
```
Comment (mattigrthr) [inline kuwala/core/backend/app/resources/export_catalog/file/download_as_csv.json]: Please stick to the wording of the designs: "With this export block, you can download the result of a transformation or data block as a CSV file."

File: kuwala/core/backend/app/resources/export_catalog/file/download_as_csv.json [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+  "macro_parameters": [
+    {
+      "id": "file_name",
+      "name": "File name",
+      "type": "text"
+    },
+    {
+      "id": "delimiter",
+      "name": "Delimiter",
+      "type": "text"
```
Comment (mattigrthr) [inline kuwala/core/backend/app/resources/export_catalog/file/download_as_csv.json]: The delimiter should have an options array. Either `,` (`id='comma')` or `;` (`id='semicolon')`.

File: kuwala/core/canvas/src/components/Modals/BlockCatalog/ExportCatalog/ExportCatalogModal.js [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+    useEffect(() => {
+        initExportCatalogs().then(null);
+    }, [])
+
+    useEffect(()=> {
+        if(catalogCategories.length) {
+            const category = catalogCategories[selectedExportIndex]
+            fetchCatalogBodyItems(category).then(null);
+        }
+    }, [selectedExportIndex])
```
Comment (mattigrthr) [inline kuwala/core/canvas/src/components/Modals/BlockCatalog/ExportCatalog/ExportCatalogModal.js]: You should to add `catalogCategories` as a dependency if you use it inside a `useEffect`.

Also all function you declare in it should be either declared inside `useEffect` and all used variables added to the dependencies, or you wrap `fetchCatalogBodyItems` with `useCallback`.

I did this refactoring to fix a bug in the `DataBlockConfigModal` on my current working branch:

``` js
    const fetchSchema = useCallback(async () => {
        if (selectedElement) {
            setIsSchemaLoading(true);

            const res = await getSchema(selectedElement.data.dataSource.id);

            if (res.status === 200) {
                const populatedSchema = populateSchema(res.data, selectedElement.data.dataSource);

                setSchema(populatedSchema)
            }

            setIsSchemaLoading(false)
        }
    }, [selectedElement])
```

Your IDE should give you warnings for this.

File: kuwala/core/canvas/src/components/Modals/BlockCatalog/TransformationCatalog/TransformationCatalogModal.js [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+                setSelectedCatalogOption={setSelectedCatalogOption}
+                selectedCatalogOption={selectedCatalogOption}
+                catalogOptions={catalogOptions}
+            />
+
+            <TransformationCatalogFooter
+                selectedCatalogOption={selectedCatalogOption}
+                toggleBlockCatalogModal={toggleBlockCatalogModal}
+                addToCanvas={async () => {
+                    console.log("Trigger add to canvas");
```
Comment (mattigrthr) [inline kuwala/core/canvas/src/components/Modals/BlockCatalog/TransformationCatalog/TransformationCatalogModal.js]: Pls remove `console.log`.

**Review by mattigrthr — CHANGES_REQUESTED** [review]
### Export block category color

For the export block categories, the color should be "purple" (see designs).

Current implementation:
image

File: kuwala/core/backend/app/main.py (hunk 9 above)
Comment (arifluthfi16) [reply kuwala/core/backend/app/main.py]: renamed to ex_dir

File: kuwala/core/canvas/src/components/Modals/BlockCatalog/ExportCatalog/ExportCatalogModal.js (hunk 13 above)
Comment (arifluthfi16) [reply kuwala/core/canvas/src/components/Modals/BlockCatalog/ExportCatalog/ExportCatalogModal.js]: refactored, moved function declaration inside useEffect.

File: kuwala/core/backend/app/main.py (hunk 8 above)
Comment (arifluthfi16) [reply kuwala/core/backend/app/main.py]: it was from existing code in line 162-170, but anyway it is refactored into `export_dir` to make it more readable

File: kuwala/core/canvas/src/components/Modals/BlockCatalog/ExportCatalog/ExportCatalogModal.js (hunk 13 above)
Comment (mattigrthr) [reply kuwala/core/canvas/src/components/Modals/BlockCatalog/ExportCatalog/ExportCatalogModal.js]: Still looks the same to me

image

File: kuwala/core/canvas/src/components/Modals/BlockCatalog/TransformationCatalog/TransformationCatalogModal.js [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+    useEffect(() => {
+        initTransformationCatalogs().then(null);
+    }, [])
+
+    useEffect(()=> {
+        if(catalogCategories.length) {
+            const category = catalogCategories[selectedTransformationIndex]
+            fetchCatalogBodyItems(category).then(null);
+        }
+    }, [selectedTransformationIndex])
```
Comment (mattigrthr) [inline kuwala/core/canvas/src/components/Modals/BlockCatalog/TransformationCatalog/TransformationCatalogModal.js]: Also here you should add `catalogCategories`.

Comment (mattigrthr) [reply kuwala/core/canvas/src/components/Modals/BlockCatalog/TransformationCatalog/TransformationCatalogModal.js]: And `fetchCatalogBodyItems` has to be either declared inside `useEffect` or with `useCallback` as mentioned in the last review.

**Review by mattigrthr — CHANGES_REQUESTED** [review]
`useEffect` doesn't seem to be refactored.

**Review by mattigrthr — APPROVED** [review]

**Pull request merged**